File: crate/src/utils/combinations_test.rs
#[cfg(test)]
mod select {
  use crate::utils::combinations::Combinations;

  #[test]
  fn it_returns_c_4_1_is_4() {
    let combinations = Combinations::new(4);                        // columns * rows
    let from = ['a', 'b', 'c', 'd'].to_vec();
    let pick = 1;
    let actual = combinations.select(&from, pick);

    println!("{:?}", actual);
    // assert!(false);
    let expected = vec![
      vec![&'a'],
      vec![&'b'],
      vec![&'c'],
      vec![&'d']
    ];

    assert_eq!(expected, actual);
  }

  #[test]
  fn it_returns_c_4_2_is_6() {                                      // C(4, 2) = 4 * 3 / 2 = 6
    let combinations = Combinations::new(4);
    let from = vec![1, 2, 3, 4];
    let pick = 2;
    let actual = combinations.select(&from, pick);

    let expected = vec![
      vec![&1, &2],
      vec![&1, &3],
      vec![&2, &3],
      vec![&1, &4],
      vec![&2, &4],
      vec![&3, &4]
    ];

    assert_eq!(expected, actual);
  }

  #[test]
  fn it_returns_c_4_3_is_4() {
    let combinations = Combinations::new(4);
    let from = ['a', 'b', 'c', 'd'].to_vec();
    let pick = 3;
    let actual = combinations.select::<char>(&from, pick);

    let expected = vec![
      vec![&'a', &'b', &'c'],
      vec![&'a', &'b', &'d'],
      vec![&'a', &'c', &'d'],
      vec![&'b', &'c', &'d']
    ];

    assert_eq!(expected, actual);
  }

  #[test]
  fn it_returns_c_4_4_is_1() {
    let combinations = Combinations::new(4);
    let from = ['a', 'b', 'c', 'd'].to_vec();
    let pick = 4;
    let actual = combinations.select::<char>(&from, pick);

    let expected = vec![vec![&'a', &'b', &'c', &'d']];

    assert_eq!(expected, actual);
  }
}

File: src/solver/utils/combinations.test.ts
import { Combinations } from "./combinations";

describe("combinations", () => {
	let combinations;//: ICombinations<string>;
	let from: string[];

	beforeEach(() => {
		combinations = new Combinations(9);                            	// Columns * rows
		from = ["a", "b", "c", "d"];
	});

	it("C(4, 1) = 4", () => {
		const pick = 1;
		const items: string[][] = combinations.select(from, pick);

		const expected: string[][] = [];
		expected[0] = ["a"];
		expected[1] = ["b"];
		expected[2] = ["c"];
		expected[3] = ["d"];

		expect(items).toEqual(expected);
	});

	// C(4, 2) = 4 * 3 / 2 = 6
	it("C(4, 2) = 6", () => {
		const pick = 2;
		const items: string[][] = combinations.select(from, pick);

		const expected: string[][] = [];
		expected[0] = ["a", "b"];
		expected[1] = ["a", "c"];
		expected[2] = ["b", "c"];
		expected[3] = ["a", "d"];
		expected[4] = ["b", "d"];
		expected[5] = ["c", "d"];

		expect(items).toEqual(expected);
	});

	it("C(4, 3) = 4", () => {
		const pick = 3;
		const items: string[][] = combinations.select(from, pick);

		const expected: string[][] = [];
		expected[0] = ["a", "b", "c"];
		expected[1] = ["a", "b", "d"];
		expected[2] = ["a", "c", "d"];
		expected[3] = ["b", "c", "d"];

		expect(items).toEqual(expected);
	});

	it("C(4, 4) = 1", () => {
		const pick = 4;
		const items: string[][] = combinations.select(from, pick);

		const expected: string[][] = [];
		expected[0] = ["a", "b", "c", "d"];

		expect(items).toEqual(expected);
	});
});